### PR TITLE
Replace pyupgrade with ruff

### PR DESCRIPTION
**Pull request by nstarman**
Signed-off-by: Nathaniel Starkman (@nstarman) <nstarkman@protonmail.com>

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by nstarman** [comment]
The ruff version is at near-parity and closing in fast to the official python version. Switching to ruff now means we wait a bit for some upgrades to our code until ruff reaches parity, but our code's not broken so no harm, no foul.

**Comment by nstarman** [comment]
Found a problem.

**Comment by nstarman** [comment]
This will be a draft until we pick up the fix for #1958

**Comment by nstarman** [comment]
Rebased on #14348. When that is merged we can replace pyupgrade with ruff!

**Comment by WilliamJamieson** [comment]
> This will be a draft until we pick up the fix for [charliermarsh/ruff#1958](https://github.com/charliermarsh/ruff/issues/1958)

This appears in version `0.0.226` of `ruff` which we will pickup in #14348

File: astropy/cosmology/flrw/lambdacdm.py [hunk 1]
```diff
@@ -263,7 +263,7 @@ def phi_z(Om0, Ok0, kappa, y1, A, z):
             phi_z2 = phi_z(self._Om0, self._Ok0, kappa, y1, A, z2)
         # Get lower-right 0<b<2 solution in Om0, Ode0 plane.
         # Fot the upper-left 0<b<2 solution the Big Bang didn't happen.
-        elif (0 < b) and (b < 2) and self._Om0 > self._Ode0:
+        elif (b > 0) and (b < 2) and self._Om0 > self._Ode0:
```
Comment (mhvk) [inline astropy/cosmology/flrw/lambdacdm.py:266]: Very random, but python supports `0 < b < 2`

File: astropy/io/votable/tests/table_test.py [hunk 2]
```diff
@@ -99,9 +99,8 @@ def test_table(tmp_path):
         ("bitarray2", {"datatype": "bit", "arraysize": "16"}),
     ]
 
-    for field, type in zip(t.fields, field_types):
-        name, d = type
-        assert field.ID == name
+    for field, (name, d) in zip(t.fields, field_types):
+        assert field.ID == name  # noqa: SIM300
```
Comment (mhvk) [inline astropy/io/votable/tests/table_test.py:103]: Why would this need a `noqa`???

File: astropy/io/votable/tree.py [hunk 3]
```diff
@@ -2123,7 +2123,7 @@ def get_ref(self):
         references.
         """
         for field in self._table._votable.iter_fields_and_params():
-            if isinstance(field, Field) and field.ID == self.ref:
+            if isinstance(field, Field) and self.ref == field.ID:
```
Comment (mhvk) [inline astropy/io/votable/tree.py:2126]: This change is in the wrong direction, and really quite illogical: mind is on field, so makes sense to have it first.

File: astropy/io/votable/tree.py [hunk 4]
```diff
@@ -2188,7 +2188,7 @@ def get_ref(self):
         references.
         """
         for param in self._table._votable.iter_fields_and_params():
-            if isinstance(param, Param) and param.ID == self.ref:
+            if isinstance(param, Param) and self.ref == param.ID:
```
Comment (mhvk) [inline astropy/io/votable/tree.py:2191]: Same, bad change.

File: astropy/modeling/separable.py [hunk 5]
```diff
@@ -142,8 +141,9 @@ def _arith_oper(left, right):
     -------
     result : ndarray
         Result from this operation.
-    """
-    # models have the same number of inputs and outputs
+    """  # noqa: D202
```
Comment (mhvk) [inline astropy/modeling/separable.py:144]: Why is a `noqa` needed here??

File: astropy/nddata/tests/test_utils.py [hunk 6]
```diff
@@ -167,7 +167,9 @@ def test_extract_array_even_shape_rounding():
     positions = (-0.99, -0.51, -0.5, -0.49, -0.01, 0)
     exp1 = (-99, 0)
     exp2 = (0, 1)
-    expected = [exp1,] * 6 + [
```
Comment (mhvk) [inline astropy/nddata/tests/test_utils.py:170]: Like done below, this should go on one line by removing the trailing `,`, which are not needed for lists.

File: astropy/stats/funcs.py [hunk 7]
```diff
@@ -1325,9 +1325,9 @@ def kuiper_false_positive_probability(D, N):
     if D < 0.0 or D > 2.0:
         raise ValueError("Must have 0<=D<=2 by definition of the Kuiper test")
 
-    if D < 2.0 / N:
+    if D < 2.0 / N:  # noqa: SIM300
```
Comment (mhvk) [inline astropy/stats/funcs.py:1328]: If `SIM300` works so badly, I think we should not use it. It is really distracting to have `noqa` in the code, so it better be meaningful.

File: astropy/stats/funcs.py [hunk 8]
```diff
@@ -1337,7 +1337,7 @@ def kuiper_false_positive_probability(D, N):
             / N ** (N - 2)
             / (b - a)
         )
-    elif (D > 0.5 and N % 2 == 0) or (D > (N - 1.0) / (2.0 * N) and N % 2 == 1):
+    elif (D > 0.5 and N % 2 == 0) or ((N - 1.0) / (2.0 * N) < D and N % 2 == 1):
```
Comment (mhvk) [inline astropy/stats/funcs.py:1340]: Another bad change.

File: astropy/units/tests/test_structured.py [hunk 9]
```diff
@@ -582,7 +582,7 @@ def test_equality(self):
         assert np.all(q1.value * u.Unit("AU, AU/day") == q_pv)
         assert not np.any(q1.value * u.Unit("AU, AU/day") != q_pv)
         assert (q_pv == "b") is False
-        assert ("b" != q_pv) is True
+        assert (q_pv != "b") is True
```
Comment (mhvk) [inline astropy/units/tests/test_structured.py:585]: Here the reversion was on purpose too.

File: astropy/visualization/lupton_rgb.py [hunk 10]
```diff
@@ -233,7 +233,7 @@ def map_intensity_to_uint8(self, I):
                 I <= 0,
                 0,
                 np.where(
-                    I >= self._range, self._uint8Max / I, self._uint8Max / self._range
+                    self._range <= I, self._uint8Max / I, self._uint8Max / self._range
```
Comment (mhvk) [inline astropy/visualization/lupton_rgb.py:236]: Bad change, given that first `np.where` is on `I` too.

File: astropy/visualization/lupton_rgb.py [hunk 11]
```diff
@@ -267,7 +267,7 @@ def __init__(self, minimum, stretch, Q=8):
             Q = 0.1
         else:
             Qmax = 1e10
-            if Q > Qmax:
+            if Qmax < Q:
```
Comment (mhvk) [inline astropy/visualization/lupton_rgb.py:270]: Super bad change.

File: astropy/wcs/tests/test_tab.py [hunk 12]
```diff
@@ -46,7 +46,7 @@ def test_2d_spatial_tab_vs_model():
 
 
 @pytest.mark.skipif(
-    _WCSLIB_VER < Version("7.6"),
+    Version("7.6") > _WCSLIB_VER,
```
Comment (mhvk) [inline astropy/wcs/tests/test_tab.py:49]: What?!?!?

File: astropy/wcs/tests/test_tabprm.py [hunk 13]
```diff
@@ -48,7 +48,7 @@ def test_tabprm_delta(tab_wcs_2di):
 
 def test_tabprm_K(tab_wcs_2di):
     t = tab_wcs_2di.wcs.tab[0]
-    assert np.all(t.K == [4, 2])
+    assert np.all([4, 2] == t.K)
```
Comment (mhvk) [inline astropy/wcs/tests/test_tabprm.py:51]: Bad change again.

**Review by mhvk — COMMENTED** [review]
I think this flipping is too often stupidly wrong here, so we should *not* turn this on.

**Comment by mhvk** [comment]
@WilliamJamieson - indeed, all the line removals etc. were fine; it was just the `SIM300` changes that went beyond maybe-not-quite-what-I-would-have-done to plain bad.  I agree that there probably is a larger discussion to be it. One possibility is not to auto-enable it, but just run it, say, once per release to have a look. More relevantly is probably to give feedback to `ruff` to make it a bit less aggressive.

Anyway, all that for later!  For this PR, I think that the `SIM300` changes make it unacceptable -- the changes are bad and adding `noqa` everywhere does not seem a good solution. Is it possible to remove this from this PR? Or does `pyupgrade` make those changes too? (I think so, since I recall looking at earlier such changes -- which all made sense!) If so, that would suggest `ruff` is not (yet) a good replacement, so then the choice would be whether we keep `pyupgrade` and not use `ruff` instead, or just disable `SIM300` for now and not worry too much about new code possibly introducing not-quite-ideally ordered comparisons.

**Comment by WilliamJamieson** [comment]
> @WilliamJamieson - indeed, all the line removals etc. were fine; it was just the `SIM300` changes that went beyond maybe-not-quite-what-I-would-have-done to plain bad. I agree that there probably is a larger discussion to be it. One possibility is not to auto-enable it, but just run it, say, once per release to have a look. More relevantly is probably to give feedback to `ruff` to make it a bit less aggressive.
> 
My feeling is that `SIM300` should probably just be ignored for astropy as a whole, or at least in the tests.

> Anyway, all that for later! For this PR, I think that the `SIM300` changes make it unacceptable -- the changes are bad and adding `noqa` everywhere does not seem a good solution. Is it possible to remove this from this PR? Or does `pyupgrade` make those changes too? (I think so, since I recall looking at earlier such changes -- which all made sense!) If so, that would suggest `ruff` is not (yet) a good replacement, so then the choice would be whether we keep `pyupgrade` and not use `ruff` instead, or just disable `SIM300` for now and not worry too much about new code possibly introducing not-quite-ideally ordered comparisons.

I think the majority of these `SIM300` issues come from @nstarman's rebase of these changes on #14348. So maybe this conversation should be moved there?

**Comment by mhvk** [comment]
@WilliamJamieson - didn't realize the `SIM300` changes were not part of this really. I now made a comment in #14348. 

@nstarman - I suggest we try to move a bit more step-by-step here! I'm now thoroughly confused whether `pyupgrade` can in fact be replaced by `ruff`, which was probably not the intent of this PR!

**Comment by nstarman** [comment]
@mhvk sorry for the confusion. This PR is still marked as a draft. It's not ready for a review since it's dependent on another PR.

**Review by mhvk — APPROVED** [review]
Looks good!

**Comment by pllim** [comment]
Thanks, all!

**Pull request merged**